File: kotlin/java-basic-example-vs-kotlin/src/main/java/org/example/ex08/Exam08.java
package org.example.ex08;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

import lombok.Data;

/**
 * @author : Rene
 * @since : 2023/10/22
 */
public class Exam08 {
	public static void main(String[] args) {

		var registerDto = new Store();

		new Exam08(registerDto);
	}

	public Exam08(Store store){
		// null point -> 터짐
		var stringRegisteredAt = toLocalDateTimeString(store.getRegisteredAt());
	}

	public String toLocalDateTimeString(LocalDateTime localDateTime){

		// null을 방지하려고 이런식으로 방어 로직을 짬
		LocalDateTime temp;
		if(localDateTime == null){
			temp = localDateTime.now();
		}

		// 혹은 optional로 받아서 orElseGet으로

		return localDateTime.format(DateTimeFormatter.ofPattern("yyyy MM dd"));
	}
}


@Data
class Store {
	private LocalDateTime registeredAt;
}



// 이러한 null 체크 로직, optional을 코틀린에서 더 쉽게 처리할 수 있다면 ?

File: kotlin/kotlin-basic-example/src/main/kotlin/ex08/Exam08.kt
package ex08

import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

/**
 * @author        : Rene
 * @since         : 2023/10/22
 */
fun main() {

    Exam08(Store())
}

data class Store(
    // data 클래스를 만들 때부터 null일 수 있음을 명시해줌
    var registeredAt: LocalDateTime ?= null

)

class Exam08{

    constructor(store: Store){
    }

    // 메서드에서 반환 -> 콜론으로 -> 뭘 리턴하는지는 마지막에 표현함
    fun toLocalDateTimeString(localDateTime: LocalDateTime?):String{

        // 이것만으로 변수 세팅이 됨
//        var temp = localDateTime ?: LocalDateTime.now()

        // ?을 작성하면 ->
//        return localDateTime?.format(DateTimeFormatter.ofPattern("yyyy MM dd"))

        // 합치면 ->
        return (localDateTime ?: LocalDateTime.now()).format(DateTimeFormatter.ofPattern("yyyy MM dd"))
    }
}